src/Psl/Type/Internal/BackedEnumValueType.php

```diff
@@ -0,0 +1,101 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Psl\Type\Internal;
+
+use BackedEnum;
+use Psl\Type\Exception\AssertException;
+use Psl\Type\Exception\CoercionException;
+use Psl\Type\Type;
+
+use function is_int;
+use function is_string;
+use function Psl\Type\int;
+use function Psl\Type\string;
+
+/**
+ * @template T of BackedEnum
+ *
+ * @extends Type<value-of<T>>
+ *
+ * @internal
+ */
+final readonly class BackedEnumValueType extends Type
+{
+    private bool $isStringBacked;
+
+    /**
+     * @psalm-mutation-free
+     *
+     * @param class-string<T> $enum
+     */
+    public function __construct(
+        private string $enum
+    ) {
+        $this->isStringBacked = is_string($this->enum::cases()[0]->value);
+    }
+
+    /**
+     * @psalm-assert-if-true value-of<T> $value
+     */
+    public function matches(mixed $value): bool
+    {
+        if (! is_string($value) && ! is_int($value)) {
```

Comment:
I think this if statement can e removed entirely, since the string/int checks are done as part of the match